revocation/ocsp/ocsp.go

```diff
@@ -0,0 +1,180 @@
+			}
+		} else {
+			if err := validateIssuer(cert, cert); err != nil {
+				return CheckOCSPError{Err: errors.New("invalid chain: expected chain to end with root cert")}
+			}
+		}
+	}
+
+	wg := sync.WaitGroup{}
+	certResults := make([]error, len(opts.CertChain))
+	for i, cert := range opts.CertChain {
+		if i != (len(opts.CertChain) - 1) {
+			wg.Add(1)
+			// Assume cert chain is accurate and next cert in chain is the issuer
+			go certOCSPStatus(cert, opts.CertChain[i+1], &certResults[i], &wg, opts)
+		} else {
+			// Last is root cert, which will never be revoked by OCSP
+			certResults[len(opts.CertChain)-1] = nil
+		}
+	}
+
+	wg.Wait()
+	return opts.MergeErrorsFunc(certResults)
+}
+
+func validateIssuer(subject *x509.Certificate, issuer *x509.Certificate) error {
+	if err := subject.CheckSignatureFrom(issuer); err != nil {
+		return err
+	}
+	if !bytes.Equal(issuer.RawSubject, subject.RawIssuer) {
+		return errors.New("parent's subject does not match issuer for a cert in the chain")
+	}
+	return nil
+}
+
+func certOCSPStatus(cert *x509.Certificate, issuer *x509.Certificate, result *error, wg *sync.WaitGroup, opts Options) {
+	defer wg.Done()
+	var err error
+
+	ocspURLs := cert.OCSPServer
+	if len(ocspURLs) == 0 {
+		// OCSP not enabled for this certificate.
+		(*result) = NoOCSPServerError{}
+		return
+	}
+
+	serverErrs := make([]error, len(ocspURLs))
+	for serverIndex, server := range ocspURLs {
+		resp, err := ocspRequest(cert, issuer, server, opts)
+		if err != nil {
+			// If there is a server error, attempt all servers before determining what to return to the user
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		if opts.SigningTime.After(resp.NextUpdate) {
+			err = errors.New("expired OCSP response")
+			serverErrs[serverIndex] = err
+			continue
+		}
+
+		foundNoCheck := false
+		for _, extension := range resp.Extensions {
+			if !foundNoCheck && extension.Id.String() == pkixNoCheckOID {
+				foundNoCheck = true
+			}
+		}
+		if !foundNoCheck {
+			// This will be ignored until CRL is implemented
+			// If it isn't found, CRL should be used to verify the OCSP response
+			// TODO: add CRL support
+			// https://github.com/notaryproject/notation-core-go/issues/125
+			serverErrs[serverIndex] = PKIXNoCheckError{}
+			continue
+		}
+
+		if resp.Status == ocsp.Revoked {
+			if time.Now().After(resp.RevokedAt) {
+				(*result) = RevokedError{}
+				return
+			} else {
+				(*result) = nil
+				return
+			}
+		} else if resp.Status == ocsp.Good {
+			(*result) = nil
+			return
+		} else {
+			(*result) = UnknownStatusError{}
+			return
+		}
+
+	}
+	err = opts.MergeErrorsFunc(serverErrs)
+	(*result) = err
+}
+
+func ocspRequest(cert, issuer *x509.Certificate, server string, opts Options) (*ocsp.Response, error) {
+	ocspRequest, err := ocsp.CreateRequest(cert, issuer, &ocsp.RequestOptions{Hash: crypto.SHA256})
```

Comment:
How is `Hash: crypto.SHA256` value determined?

Reply:
This was previously crypto.SHA1, but I changed it to crypto.SHA256 per your suggestion. Doing some additional research, I see that this is fairly standard. There is even this documentation from Microsoft: https://learn.microsoft.com/en-us/azure/security/fundamentals/ocsp-sha-1-sunset

Reply:
Is there a way to determine this value from the certificate itself so that we can avoid hardcoding?

Reply:
Unfortunately, this field does not exist. It is up to the requester to determine what type of Hash they would like the response to be in. Based on the above specification, I believe SHA256 is the best approach.

Reply:
If it is upto the requester to determine, why are we hardcoding it to SHA256? Why not SHA384 for example? I feel  it is better to use certificate's hashing alg for this so that the code automatically uses higher algs as the certificates themselves upgrade.

Reply:
The choice for SHA256 rather than a different hashing algorithm, SHA384 for example, was based upon the [Microsoft specification linked above](https://learn.microsoft.com/en-us/azure/security/fundamentals/ocsp-sha-1-sunset) that specified "All OCSP responses will use the SHA-256 hashing algorithm." Thus, I assumed this is the best choice currently.

---

W.r.t modifying the algorithm based on the certificate, we cannot directly obtain this hash algorithm value from the certificate itself. The options for specifying a Hash (as described in the documentation [here](https://pkg.go.dev/golang.org/x/crypto/ocsp#RequestOptions) must be one of [these types](https://pkg.go.dev/crypto#Hash). However, the [certificate](https://pkg.go.dev/crypto/x509#Certificate) only specifies a [SignatureAlgorithm](https://pkg.go.dev/crypto/x509#SignatureAlgorithm) and a [PublicKeyAlgorithm](https://pkg.go.dev/crypto/x509#PublicKeyAlgorithm). The only way we could potentially derive this is by using a switch with the SignatureAlgorithm and converting it on our end into one of the hashes.

However, there are some signature algorithms that do not have one of those hashes associated with them. Additionally, some of the signature algorithms specify SHA1 for example, which should no longer be used per the [above documentation](https://learn.microsoft.com/en-us/azure/security/fundamentals/ocsp-sha-1-sunset). For these algorithms we would still need to specify a default hash to use (which would need to be upgraded over time anyways).

Additionally, as new signature algorithms and hashes are added, we would still need to make modifications to adjust what SignatureAlgorithms have hash types that are no longer valid, what new algorithms correspond with a specific hash, etc.

TLDR: I think we will need to upgrade this value/logic regardless over time, so I'd argue it makes sense to keep it consistent rather than complicate the process. This would ensure that users can more easily understand what algorithm is being used. Maybe an alternative would be to have a default (SHA256) and allow users to pass a different hash type if they'd prefer?

Reply:
Doing more research into [the RFC](https://datatracker.ietf.org/doc/rfc6960/), this extension for a preferred algorithm seems to only be a best intention. Ultimately, the responder can choose whatever supported algorithm it wants (with some restrictions). However, there is a requirement to support SHA256 and SHA1 in the specification. Additionally, the client is required to support these two algorithms in the specification, as well as any preferred algorithm that is attached to the request. With this in mind, we would need to pass either SHA256 or SHA1 anyways as a backup option if the preferred method is not supported or not chosen.

As a result, I think it is a poor user experience for them to request a different algorithm and not have it be used. The alternative method of retrieving the algorithm from the cert may work, but we would still need to default to SHA256 (and provide it as a backup).

Based on the specification, do you still think modifying the requested algorithm makes sense? I believe it would make sense to wait until higher algorithms are required for all OCSP responders

Reply:
> Additionally, the client is required to support these two algorithms in the specification, as well as any preferred algorithm that is attached to the request. With this in mind, we would need to pass either SHA256 or SHA1 anyways as a backup option if the preferred method is not supported or not chosen.

Can you clarify what happens when the client requested alg is not supported by the server? Does the server outright fail the request or return a response with it's own alg?

Reply:
If the requested algorithm is not supported, it will fall back to its own algorithm. To clarify, there's actually a series of recommended steps for how the responder should choose what algorithm to use. However, it is not enforced for it to choose what is preferred. As long as the server responds with an algorithm that the client is required to support. Essentially, the brute of the algorithm support is on the client not the server. The client can request one, but the server does not need to obey.

Reply:
After syncing up and discussing the options, the consensus was to default to SHA256 for now and track this in a new issue: https://github.com/notaryproject/notation-core-go/issues/139